src/colmap/util/threading_ctrl.h

```diff
@@ -0,0 +1,270 @@
+#include <climits>
+#include <functional>
+#include <future>
+#include <list>
+#include <queue>
+#include <thread>
+#include <type_traits>
+#include <unordered_map>
+
+namespace colmap {
+
+#ifdef __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wkeyword-macro"
+#endif
+
+#ifdef __clang__
+#pragma clang diagnostic pop  // -Wkeyword-macro
+#endif
+
+// Reimplementation of threading with thread-related functions outside
+// controller Following util/threading.h
+
+// Core methods of a controller, wrapped by BaseController
+class CoreController {
+ public:
+  CoreController(){};
+  virtual ~CoreController() = default;
+
+  // Set callbacks that can be triggered within the main run function.
+  void AddCallback(int id, const std::function<void()>& func);
+
+  // This is the main run function to be implemented by the child class. If you
+  // are looping over data and want to support the pause operation, call
+  // `BlockIfPaused` at appropriate places in the loop. To support the stop
+  // operation, check the `IsStopped` state and early return from this method.
+  virtual void Run() = 0;
+
+ protected:
+  // Register a new callback. Note that only registered callbacks can be
+  // set/reset and called from within the thread. Hence, this method should be
+  // called from the derived thread constructor.
+  void RegisterCallback(int id);
+
+  // Call back to the function with the specified name, if it exists.
+  void Callback(int id) const;
+
+ private:
+  std::unordered_map<int, std::list<std::function<void()>>> callbacks_;
+};
+
+// BaseController that supports templating in ControllerThread
+class BaseController : public CoreController {
+ public:
+  struct ThreadStatus {
+    // Check the status
+    bool IsStarted() { return started; }
+    bool IsStopped() { return stopped; }
+    bool IsPaused() { return paused; }
+    bool IsRunning() { return started && !pausing && !finished; }
+    bool IsFinished() { return finished; }
+
+    // Update the status
+    void Start() {
+      started = true;
+      stopped = false;
+      paused = false;
+      pausing = false;
+      finished = false;
+      setup = false;
+      setup_valid = false;
+    }
+
+    bool started = false;
+    bool stopped = false;
+    bool paused = false;
+    bool pausing = false;
+    bool finished = false;
+    bool setup = false;
+    bool setup_valid = false;
+  };
+
+  enum {
+    STARTED_CALLBACK = INT_MIN,
+    FINISHED_CALLBACK,
+
+    // thread-related callbacks from here, only useful in the Thread class,
+    // empty as default
+    LOCK_MUTEX_CALLBACK,
+    BLOCK_IF_PAUSED_CALLBACK,
+    SIGNAL_SETUP_CALLBACK,
+    CHECK_VALID_SETUP_CALLBACK,
+  };
+
+  BaseController();
+  virtual ~BaseController() = default;
+
+  ///////////////////////////////////////////////////
+  // Thread-related functions
+  ///////////////////////////////////////////////////
+  ThreadStatus status_;
+  ThreadStatus* GetThreadStatus() { return &status_; }
```

Comment:
return const ref to avoid manipulation outside?

Reply:
This function is removed with the new update, where all the thread status goes into ControllerThread, where they are made private.